### ISSUE TITLE
[RFC]: Stage-local speculative decoding for Qwen3-Omni: Thinker first, then Talker

### ISSUE THREAD

**Issue by amy-why-3459** [vllm-project/vllm-omni#7837]
## Motivation

Introduce independently configurable speculative decoding for the Qwen3-Omni Thinker and Talker in vLLM-Omni, with two delivery phases:

1. **Phase 1: speculative Thinker, ordinary Talker**, progressing from text-only output to full-payload speech and then async-chunk speech.
2. **Phase 2: speculative Talker**, culminating in both stages running speculation concurrently without exposing unverified text conditions or audio.

The objective is lower end-to-end generation latency while preserving the target model's sampling semantics and the consistency of inter-stage conditioning. Faster isolated text decoding is insufficient if conditioning rows are dropped, audio is misaligned, or streaming latency regresses.

This is a design proposal, not a claim that existing checkpoints/configurations already support these combinations. No acceleration factor is claimed in advance.

### Repository baseline and concrete integration points

This proposal was checked against vLLM-Omni main at [`f90c267`](https://github.com/vllm-project/vllm-omni/commit/f90c267ffb61c79e42f0165476909b995caedd83). Implementation PRs should additionally pin their vLLM commit, model revision, hardware, attention backend, and sampling configuration.

- [Qwen3-Omni pipeline](https://github.com/vllm-project/vllm-omni/blob/f90c267ffb61c79e42f0165476909b995caedd83/vllm_omni/model_executor/models/qwen3_omni/pipeline.py): stage 0 is the `LLM_AR` Thinker, stage 1 the `LLM_AR` Talker, and stage 2 `LLM_GENERATION` Code2Wav. A Thinker-only pipeline also exists. Keep this topology; Code2Wav is outside the speculation scope.
- [GPUARModelRunner](https://github.com/vllm-project/vllm-omni/blob/f90c267ffb61c79e42f0165476909b995caedd83/vllm_omni/worker/gpu_ar_model_runner.py): already has speculative proposal/sampling plumbing. Audit its override against the pinned upstream implementation instead of building a second scheduler. `_should_use_async_omni_output()` currently returns false when speculation is configured: **async scheduling, async-chunk transport, and async output materialization are distinct capabilities**.
- [Qwen3 stage processors](https://github.com/vllm-project/vllm-omni/blob/f90c267ffb61c79e42f0165476909b995caedd83/vllm_omni/model_executor/stage_input_processors/qwen3_omni.py): `thinker2talker_async_chunk` rejects multiple post-prefill embeddings for ordinary non-resumable requests. Resumable requests follow a different helper and need separate validation. Also cover `thinker2talker_full_payload`, `thinker2talker_token_only`, and both Talker-to-Code2Wav producers.
- [Qwen3 model wrapper](https://github.com/vllm-project/vllm-omni/blob/f90c267ffb61c79e42f0165476909b995caedd83/vllm_omni/model_executor/models/qwen3_omni/qwen3_omni.py): `talker_postprocess` retains the last hidden row; `talker_preprocess` tracks text consumption and prepares `mtp_inputs`; `talker_mtp` obtains full codec codes and **summed codec embeddings** from the code predictor and combines those embeddings with `text_step`. These are stateful dependencies that a multi-step verifier must respect.
- [Talker](https://github.com/vllm-project/vllm-omni/blob/f90c267ffb61c79e42f0165476909b995caedd83/vllm_omni/model_executor/models/qwen3_omni/qwen3_omni_moe_talker.py): `compute_logits` predicts RVQ codebook 0; the code predictor generates the remaining codebooks. This internal MTP/code-predictor operation is not automatically vLLM speculative `method="mtp"`.
- [ChunkTransferAdapter](https://github.com/vllm-project/vllm-omni/blob/f90c267ffb61c79e42f0165476909b995caedd83/vllm_omni/distributed/omni_connectors/transfer_adapter/chunk_transfer_adapter.py): already carries chunk counters, sender-generation fences, and captured `confirmed_num_computed_tokens`/`segment_generation`. Extend these mechanisms rather than adding a parallel transport protocol.

The runner's optional `_run_post_sample_talker_mtp` hook rejects speculative token rows, but this must **not** be presented as proof that Qwen3 uses that hook. Qwen3's actual preprocess/MTP/postprocess path above is the integration target.

Related: [#5348](https://github.com/vllm-project/vllm-omni/issues/5348) documents a historical `ngram_gpu`/async-bookkeeping failure and the Thinker-to-Talker row-alignment concern. Reproduce or close each gap on the pinned implementation; fixing the proposer dispatch alone does not establish audio correctness.

### Goals and initial scope

- Stage-local enablement and independent draft lengths, with speculation disabled by default.
- Preserve accepted token/frame order, target sampling semantics, KV consistency, and final-flush behavior.
- Publish an explicit model/method/backend/concurrency support matrix; reject unsupported combinations early.
- Initially target ordinary turn-mode generation. Resumable/duplex input, barge-in, and KV-window reanchoring remain disabled with speculation until separately validated.
- Do not speculate Code2Wav or send provisional audio. Do not promise simultaneous support for every proposer, quantization, TP/PP configuration, or accelerator.

### Commit invariants

1. Only the target verifier may commit outputs. No rejected/provisional draft row reaches a downstream stage.
2. A sampled token, its computed KV, and its downstream conditioning need not become available simultaneously. Track these boundaries explicitly.
3. Published conditions and playable audio are irreversible. Rejection rolls back local provisional state only.
4. A logical position is submitted once and in order, including after cancellation, scheduling changes, and transport retries.
5. A terminal marker follows the last valid committed payload. Request/segment completion must retain the existing distinction.

```text
Committed multimodal input
    -> Thinker draft / target verification
    -> accepted token + ready-condition commit
    -> existing connector
    -> ordinary Talker (Phase 1) / Talker draft + verification (Phase 2)
    -> committed complete codec frames
    -> Code2Wav -> audio
```

There is no cross-stage speculative dependency: Talker consumes only committed Thinker conditions. The stages may overlap normally, but rejected Thinker drafts never require downstream rollback.

#### 1A. Establish the text-only baseline

Use a validated Thinker-only setup to audit proposal, target verification, rejection sampling, accepted counts, KV recovery, and bookkeeping. Select one proposer/backend combination first. CPU n-gram can serve as a low-integration-cost correctness harness, not a promise of useful acceptance rates or learned-drafter compatibility.

A learned drafter requires an explicitly supported checkpoint, tokenizer/vocabulary compatibility, hidden-state interface, positional encoding, and multimodal context handling. Do not assume Qwen3-VL or text-Qwen3 draft weights are interchangeable with Qwen3-Omni Thinker weights. Dedicated upstream configuration validation is not a substitute for an available compatible checkpoint and end-to-end tests.

Preserve upstream proposer-specific timing: GPU-token proposers and CPU-token proposers cannot automatically share the same post-bookkeeping path. Test empty sampled rows, no draft, all-rejected, partially accepted, all-accepted, EOS, and multiple requests with different acceptance lengths.

#### 1B. Add acceptance-aware condition materialization

Introduce a small model-side adapter consuming the generic verifier's acceptance/index information and producing Qwen3-Omni-compatible conditioning.

Audit the actual model contract for each tensor: input embedding, hidden state predicting a token, or hidden state after consuming that token. Build an explicit token-position-to-target-row mapping rather than selecting `hidden_states[:accepted_count]`.

**Replacement and bonus tokens require special treatment.** A token sampled at a rejected position, or the additional token sampled after all drafts are accepted, may not yet have all downstream-required conditioning. Keep it pending until the required target computation completes. If an extra forward is necessary, include that cost in benchmarks. Conversely, do not delay a tensor whose contract makes it already available.

The logical commit record should identify request/segment generation, condition-sequence start, committed token IDs, available model-specific conditions, and their row mapping. It is an internal contract, not a mandatory new serialization format. Reuse existing `OmniPayloadStruct` fields and connector metadata where possible; large tensors stay on the connector data plane.

EOS and other special tokens follow the existing model protocol, not an unconditional one-token/one-audio-condition rule.

#### 1C. Integrate full-payload and async-chunk speech

First validate Thinker speculation with ordinary Talker/Code2Wav via full-payload transfer. Then extend async-chunk transport:

- Support a variable-length committed decode span, either consumed as a span or split into ordered rows at the model adapter.
- Distinguish input-prefill segments from accepted multi-token decode spans explicitly. Tensor row count alone is insufficient.
- Align token IDs, embeddings, hidden rows, and captured confirmed counts in the same snapshot before enqueueing asynchronous work.
- Reuse sender-generation fences and chunk ordering. Detect gaps/duplicates; discard stale work after cancellation.
- Update placeholder/reservation accounting to reflect the actual downstream model input, including pending conditions and final flush.

Keep `_should_use_async_omni_output()`'s speculation guard until accepted-row snapshots and deferred materialization are implemented and tested. Async-chunk speech may initially use the synchronous materialization path; restoring the fast path is a separate optimization, not a prerequisite to claiming transport support.

**Phase 1 completion:** Thinker speculation with ordinary Talker passes text-only, full-payload speech, and ordinary async-chunk speech acceptance criteria on the declared support matrix. Completing text-only work is a useful submilestone but is not completion of Phase 1.

#### 2A. Define the frame distribution and verification algorithm

Treat a complete RVQ frame as the external commit unit:

`frame_t = [code_0, code_1, ..., code_(Q-1)]`.

The Talker backbone predicts codebook 0; the code predictor generates the remaining codes. Their summed embeddings feed subsequent backbone execution. Therefore, accepting only codebook 0 does not validate a drafted complete frame or the inputs used to verify later frames.

Before integration, document:

- The target distribution over each frame and the actual proposal distribution.
- How candidate full-frame embeddings are constructed for target verification.
- Whether verification is frame-joint, sequential by codebook, or another equivalent factorization.
- How a rejected code/frame invalidates all dependent future candidate states.
- The residual sampling procedure for stochastic speculative sampling, not just an acceptance ratio.

The recommended initial prototype uses complete-frame proposals and commit boundaries, but the probability factorization remains an algorithm design gate. A frame-level `min(1,p/q)` expression alone is not an implementable sampler over the large joint codec space. Sequential verification may be correct but erase the intended speedup; measure it before committing to an implementation.

A greedy-only milestone is acceptable if clearly labeled and unsupported stochastic configurations are rejected. Do not claim distribution-preserving random sampling until the full rejection/residual procedure is implemented and tested.

#### 2B. Make Qwen3 Talker state provisional until commit

Refactor the actual `talker_preprocess`, `talker_mtp`, and `talker_postprocess` path to expose per-candidate state instead of retaining only the final hidden row or immediately mutating committed metadata.

Account for backbone KV, drafter KV, code-predictor state where persistent, last hidden state, summed codec embeddings, text-consumption counters (`num_processed_tokens` and prefill handoff), decode flags, and EOS/frame position. Define which values are recomputable and which must be snapshotted/restored.

Only complete committed frames may enter `talker2code2wav_full_payload` or `talker2code2wav_async_chunk`. Audit their frame extraction and buffering for multiple accepted frames; no provisional frame may update Code2Wav state or produce audio.

#### 2C. Run both stages together

Maintain independent draft budgets and acceptance metrics. Talker drafts only against committed, available Thinker conditioning. When more conditioning arrives, determine validity using the real text-consumption schedule; invalidate only candidates whose conditioning assumptions changed.

Use bounded queues/backpressure. Keep speculation length separate from transport/audio chunk length, so waiting to fill a draft window cannot indefinitely postpone playable audio. Benchmark four combinations: neither stage, Thinker only, Talker only, and both stages.

### Ownership and configuration

| Component | Proposed responsibility |
|---|---|
| Stage deployment/model capability validation | Route per-stage speculative configuration through existing stage engine arguments; validate actual method/model/backend/streaming combinations |
| Shared scheduler and runner | Reuse upstream proposal/verification/KV contracts; expose thin acceptance/materialization hooks without model-specific codec logic |
| Qwen3 model adapter | Thinker condition-row mapping; Talker frame verification and state commit/rollback |
| Stage processors and connector | Existing payload schema, ordered committed spans, generation fencing, cancellation and final flush |
| Orchestrator | Existing stage routing and backpressure; no GPU KV mutation or codec probability logic |

Keep the existing stage IDs and pipeline definitions. Phase 1 permits Thinker speculation and rejects Talker speculation; Phase 2 permits each independently. Do not reuse internal Talker “MTP” naming as a speculative method declaration. Add executable YAML examples only after the proposer and checkpoint are selected and tested.

This support is not inherently duplex-specific. Model-specific logic belongs with Qwen3-Omni; generic speculative infrastructure stays generic. Duplex-only interruption/reanchor integration should be isolated and validated separately.

### Tests and acceptance criteria

**Production-path integration:** extend runner tests, Qwen3 model/processor tests, connector tests, and online serving tests. Do not rely mainly on mocks that copy the production algorithm. Exercise the full schedule -> verify -> materialize -> enqueue -> consume path.

Required cases:

- No/all/partial draft acceptance; bonus/replacement tokens; EOS inside a candidate span; empty sampled rows.
- Accepted IDs, condition rows, effective KV positions, text counters, and final markers agree.
- Multiple requests with different acceptance lengths; first/chunked prefill; async scheduling; cancellation; cleanup; supported preemption/re-admission behavior.
- Text, audio, image, and video inputs; short/long outputs; full-payload and async-chunk speech.
- Talker multi-codebook rejection, dependent frame invalidation, complete-frame commit and final codec flush.
- Resumable/duplex cases only when enabling that capability: append, barge-in, generation invalidation, and KV-window shifts.

For greedy decoding, compare token/frame sequences under controlled deterministic conditions and account explicitly for numerical differences. For stochastic sampling, test the algorithm and distribution, rather than requiring identical same-seed sequences across different RNG consumption patterns. Commit synthetic small-distribution tests in addition to model-level evaluation.

**Performance:** measure proposer and verifier costs, accepted length, per-stage throughput, TTFT/TPOT, first-audio latency, audio real-time factor, chunk interval p50/p95/p99, underruns, queue time, peak memory and transport overhead. Include short/long outputs and target concurrency.

Because speculation disables the current async output fast path, compare against the **best ordinary production baseline**, not only against a baseline with that optimization disabled. Also run a controlled same-materialization-path comparison to isolate algorithmic gains. Acceptance rate alone is not the success criterion.

Pin hardware, commits, checkpoint revisions, sampling, warmup, and repeated-run methodology. Agree on workload-specific speedup/tail-latency thresholds before performance sign-off; keep default-off when there is no demonstrated end-to-end gain.

### Milestones

| Milestone | Deliverable / exit criterion |
|---|---|
| M0 | Pin baseline and support matrix; specify token/condition/frame timing |
| M1 | Thinker-only verifier, proposer timing and KV correctness |
| M2 | Accepted-condition materialization and full-payload speech |
| M3 | Ordinary async-chunk speech, cleanup and performance validation; Phase 1 complete |
| M4 | Talker distribution/proposal design and measured prototype |
| M5 | Talker frame commit/rollback and isolated correctness/performance |
| M6 | Both stages enabled; four-way benchmark and streaming acceptance; Phase 2 complete |

### Alternatives and tradeoffs

- **Only accelerate Thinker:** useful first delivery with lower implementation cost; may offer limited speech benefit when Talker dominates.
- **Send Thinker drafts downstream:** potentially more overlap, but introduces cascading rollback and irreversible audio hazards; excluded.
- **Speculate codebook 0 only:** lower proposal complexity, but requires a correct treatment of remaining codebooks and their recurrent embeddings; not sufficient by itself.
- **One global speculative controller:** couples unrelated stage lifecycles and budgets; prefer independent controllers with committed handoff.

### Questions for maintainers

1. Which Thinker proposer/checkpoint and backend should be the first supported combination?
2. Can accepted target-row metadata be exposed through existing upstream hooks, minimizing Omni runner divergence?
3. Which conditioning tensors require a follow-up target forward for bonus/replacement tokens?
4. Is a greedy-only Talker milestone acceptable, and which proposal/verification factorization is practical?
5. What are the initial hardware/parallelism support matrix and first-audio/tail-latency acceptance thresholds?
6. Should resumable/duplex support be a follow-up after ordinary async-chunk speech?

## Feedback Period

Proposed: at least two weeks after publication, with the Phase 2 algorithm reviewed before implementation begins. Timing can be adjusted by maintainers.

## CC List

Feedback welcome from Qwen3-Omni model, AR runner, connector/streaming, and speculative decoding maintainers. No individual assignees are proposed.

## Any Other Things

This RFC proposes future work; no model-level benchmark or end-to-end validation is claimed here. Repository source and related issues were inspected. The external detailed-design Google Docs template linked by the issue form was inaccessible during preparation, so the detailed design is included inline under the repository's RFC headings. The documentation chatbot was not consulted; this proposal does not claim that checklist item was completed.

**Comment by zqc214** [comment]
I'd like to take this issue, please assign it to me.
